### PR TITLE
[buildImage] Add support of platform.json parsing to portconfig.py file

### PR DESCRIPTION
**Pull request by samaity**
Signed-off-by: Sangita Maity <sangitamaity0211@gmail.com>

* [sonic-cfggen] Add a unit-test to test platform capability file

Signed-off-by: Sangita Maity <sangitamaity0211@gmail.com>

Adding a new function to create BRKOUT_CFG TABLE in config db

Signed-off-by: Sangita Maity <sangitamaity0211@gmail.com>

**- What I did**
Add support of **platform.json** parsing to **portconfig.py** file which is being used by **_sonic-cfggen_** and ***minigraph.py*** file under ***src/sonic-config-engine*** folder to get port config via get_port_config  function.

**- How I did it**

1.  **portconfig.py** file will first check whether the **platform.json** file is there or not. if not then whether port_config.ini file is there or not. Modified **get_port_config_file_name** for this purpose.
2. Added two separate functions i.e. **parse_platform_json_file**  to get port attributes from **platform.json** and **gen_port_config** to generate port attributes.
3. Added another two functions i.e  get_breakout_mode parse_breakout_mode to get breakout mode and parse breakout mode from platform.json respectively.

**- How to verify it**

rebuilt "sonic_config_engine-1.0" wheel package with all the test cases.All the below-mentioned test cases passed.

```
# Check whether all interfaces present or not as per platform.json
def test_platform_json_interfaces_keys(self):
       
# Check specific Interface with it's proper configuration as per platform.json
def test_platform_json_specific_ethernet_interfaces(self):

# Check all Interface with it's proper configuration as per platform.json
def test_platform_json_all_ethernet_interfaces(self):
```

log file
```
samaity@50c03d616603:/sonic$ BLDENV=stretch make -f slave.mk target/python-wheels/sonic_config_engine-1.0-py2-none-any.whl
SONiC Build System

Build Configuration
"CONFIGURED_PLATFORM"             : "vs"
"CONFIGURED_ARCH"                 : "amd64"
"SONIC_CONFIG_PRINT_DEPENDENCIES" : ""
"SONIC_BUILD_JOBS"                : "1"
"SONIC_CONFIG_MAKE_JOBS"          : "24"
"SONIC_USE_DOCKER_BUILDKIT"       : ""
"USERNAME"                        : "admin"
"PASSWORD"                        : "admin123"
"ENABLE_DHCP_GRAPH_SERVICE"       : ""
"SHUTDOWN_BGP_ON_START"           : ""
"ENABLE_PFCWD_ON_START"           : ""
"INSTALL_DEBUG_TOOLS"             : ""
"ROUTING_STACK"                   : "frr"
"FRR_USER_UID"                    : "300"
"FRR_USER_GID"                    : "300"
"ENABLE_SYNCD_RPC"                : ""
"ENABLE_ORGANIZATION_EXTENSIONS"  : "y"
"HTTP_PROXY"                      : ""
"HTTPS_PROXY"                     : ""
"ENABLE_SYSTEM_TELEMETRY"         : "y"
"SONIC_DEBUGGING_ON"              : ""
"SONIC_PROFILING_ON"              : ""
"KERNEL_PROCURE_METHOD"           : "build"
"BUILD_TIMESTAMP"                 : ""
"BLDENV"                          : "stretch"
"VS_PREPARE_MEM"                  : "yes"
"ENABLE_SFLOW"                    : "y"
```
```

samaity@server09:~/REPO_FACTORY/GIT_REPO/DPB/sonic-buildimage$ ls -la     target/python-wheels/sonic_config_engine-1.0-py2-none-any.whl
-rw-r--r-- 1 samaity samaity 44514 Dec  2 19:05 target/python-wheels/sonic_config_engine-1.0-py2-none-any.whl
```

### CODE REVIEW

File: src/sonic-daemon-base/sonic_daemon_base/daemon_base.py [hunk 1]
```diff
@@ -151,13 +152,17 @@ def get_path_to_port_config_file(self):
...
+        port_config_file_path = "/".join([hwsku_path, PLATFORM_CAP])
         if not os.path.isfile(port_config_file_path):
-            # port_config.ini doesn't exist. Try loading the legacy 'portmap.ini' file
-            port_config_file_path = "/".join([hwsku_path, PORTMAP])
+
+            # platform.json doesn't exist. Try loading the legacy 'port_config.ini' file
+            port_config_file_path = "/".join([hwsku_path, PORT_CONFIG])
             if not os.path.isfile(port_config_file_path):
-                raise IOError("Failed to detect port config file: %s" % (port_config_file_path))
+                # port_config.ini doesn't exist. Try loading the legacy 'portmap.ini' file
```
Comment (jleveque) [inline src/sonic-daemon-base/sonic_daemon_base/daemon_base.py]: The legacy 'portmap.ini' file has been deprecated for over 3 years at this point. I think it is safe to remove all code pertaining to this file.

@lguohan: do you agree? If so, should that change be in a separate PR, or do you think it's OK to be included in this one?

**Comment by stevenlu99** [comment]
I think would be better if we put all port definitions in platform.json file into a section "interfaces" or "port_defs"?

**Comment by samaity** [comment]
> I think would be better if we put all port definitions in platform.json file into a section "interfaces" or "port_defs"?

#done

File: src/sonic-daemon-base/sonic_daemon_base/daemon_base.py (hunk 1 above)
Comment (lguohan) [reply src/sonic-daemon-base/sonic_daemon_base/daemon_base.py:162]: i agree.

Comment (samaity) [reply src/sonic-daemon-base/sonic_daemon_base/daemon_base.py:162]: Should I remove it in this PR?

Comment (jleveque) [reply src/sonic-daemon-base/sonic_daemon_base/daemon_base.py:162]: Yes, please.

Comment (samaity) [reply src/sonic-daemon-base/sonic_daemon_base/daemon_base.py]: #done

File: src/sonic-config-engine/tests/test_cfggen_platformJson.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
+from unittest import TestCase
+import subprocess
+import os
+import json
+import ast
+
+# Global Variable
+PLATFORM_CAP = "platform_output.json"
```
Comment (jleveque) [inline src/sonic-config-engine/tests/test_cfggen_platformJson.py:8]: This is unused, yet you refer to `platform_output.json` explicitly on line 69.

Also, `PLATFORM_CAP` is not a clear name for this constant. Suggest to rename `PLATFORM_CAPABILITIES_FILE`, `PLATFORM_OUTPUT_FILE` or similar.

File: src/sonic-daemon-base/sonic_daemon_base/daemon_base.py [hunk 3]
```diff
@@ -27,7 +27,7 @@
 
 # Port config information
 PORT_CONFIG = 'port_config.ini'
-PORTMAP = 'portmap.ini'
+PLATFORM_CAP = 'platform.json'
```
Comment (jleveque) [inline src/sonic-daemon-base/sonic_daemon_base/daemon_base.py:30]: In this PR and other related PRs, you name this constant, `PLATFORM_JSON`. I would prefer if you used that everywhere, and you can change it everywhere in the future if desired.

File: src/sonic-config-engine/tests/test_cfggen_platformJson.py (hunk 2 above)
Comment (samaity) [reply src/sonic-config-engine/tests/test_cfggen_platformJson.py]: #done

File: src/sonic-daemon-base/sonic_daemon_base/daemon_base.py (hunk 3 above)
Comment (samaity) [reply src/sonic-daemon-base/sonic_daemon_base/daemon_base.py]: #done

**Comment by jleveque** [comment]
Please fix LTGM alert. Looks like integer division. If you don't care about the remainder, you can use the Python "floor division operator: `//`

**Comment by samaity** [comment]
> Please fix LGTM alert. Looks like integer division. If you don't care about the remainder, you can use the Python "floor division" operator: `//`

#done

**Comment by lguohan** [comment]
retest vsimage please

**Comment by daall** [comment]
retest vsimage please

**Comment by renukamanavalan** [comment]
retest vsimage please

**Comment by zhenggen-xu** [comment]
retest vsimage please

File: src/sonic-config-engine/portconfig.py [hunk 4]
```diff
@@ -1,33 +1,141 @@
...
+
+def db_connect_configdb():
+    """
+    Connect to configdb
+
+    config_db = ConfigDBConnector()
+    if config_db is None:
+        return None
+    try:
+        config_db.connect()
```
Comment (zhenggen-xu) [inline src/sonic-config-engine/portconfig.py:56]: This could be blocking during the config load_minigraph phase, in case the CONFIG_DB_INITIALIZED is not yet set in the configDB.  We can ignore the check by using config_db.db_connect('CONFIG_DB') instead.

**Comment by zhenggen-xu** [comment]
retest vsimage please

**Comment by daall** [comment]
retest vsimage please

**Comment by daall** [comment]
retest vsimage please

File: src/sonic-config-engine/minigraph.py [hunk 5]
```diff
@@ -772,7 +772,7 @@ def enable_internal_bgp_session(bgp_sessions, filename, asic_name):
 # Main functions
 #
 ###############################################################################
-def parse_xml(filename, platform=None, port_config_file=None, asic_name=None):
+def parse_xml(filename, platform=None, port_config_file=None, hwsku_config_file=None, asic_name=None):
```
Comment (renukamanavalan) [inline src/sonic-config-engine/minigraph.py:775]: Can we introduce the new optional arg as last one ?
This will help in scenarios, where callers pass optional args by position w/o qualifying with name.

File: src/sonic-config-engine/minigraph.py [hunk 6]
```diff
@@ -836,7 +837,10 @@ def parse_xml(filename, platform=None, port_config_file=None, asic_name=None):
         if child.tag == str(docker_routing_config_mode_qn):
             docker_routing_config_mode = child.text
 
-    (ports, alias_map, alias_asic_map) = get_port_config(hwsku=hwsku, platform=platform, port_config_file=port_config_file, asic=asic_id)
+    if hwsku_config_file:
+        (ports, alias_map, alias_asic_map) = get_port_config(hwsku=hwsku, platform=platform, port_config_file=port_config_file, hwsku_config_file=hwsku_config_file, asic=asic_id)
```
Comment (renukamanavalan) [inline src/sonic-config-engine/minigraph.py:841]: We don't really need an if/else. hwsku will be passes as None, if not there.

File: src/sonic-config-engine/portconfig.py [hunk 7]
```diff
@@ -1,33 +1,138 @@
...
+
+def db_connect_configdb():
+    """
+    Connect to configdb
+    """
+    config_db = ConfigDBConnector()
+    if config_db is None:
+        return None
+    try:
+        config_db.db_connect('CONFIG_DB')
```
Comment (renukamanavalan) [inline src/sonic-config-engine/portconfig.py:56]: If you would like, we could make the async nature of this call, explicit. Even a comment, as Xu explained, could be helpful for future code readers, to help understand, why async mode is selected.

config_db.connect(wait_for_init=False)

Xu's comments that could be re-phrased
"_This could be blocking during the config load_minigraph phase, as the CONFIG_DB_INITIALIZED is not yet set in the configDB. We can ignore the check by using config_db.db_connect('CONFIG_DB') instead_"

File: src/sonic-config-engine/minigraph.py (hunk 5 above)
Comment (samaity) [reply src/sonic-config-engine/minigraph.py]: Modified as requested!

File: src/sonic-config-engine/portconfig.py (hunk 7 above)
Comment (samaity) [reply src/sonic-config-engine/portconfig.py]: Modified as requested!

File: src/sonic-config-engine/minigraph.py (hunk 6 above)
Comment (samaity) [reply src/sonic-config-engine/minigraph.py]: Yes, correct! 👍  Modified as requested!

**Comment by samaity** [comment]
All test cases are passing except [dhcp_relay.test_dhcp_relay.test_dhcp_relay_after_link_flap](https://sonic-jenkins.westus2.cloudapp.azure.com/job/vs/job/buildimage-vs-image-pr/4321/testReport/junit/dhcp_relay/test_dhcp_relay/test_dhcp_relay_after_link_flap_2/)


I can see some ongoing [issue](https://github.com/Azure/sonic-mgmt/issues/1699) related to this matter? Any Idea?

**Comment by daall** [comment]
retest vsimage please

**Comment by daall** [comment]
> All test cases are passing except [dhcp_relay.test_dhcp_relay.test_dhcp_relay_after_link_flap](https://sonic-jenkins.westus2.cloudapp.azure.com/job/vs/job/buildimage-vs-image-pr/4321/testReport/junit/dhcp_relay/test_dhcp_relay/test_dhcp_relay_after_link_flap_2/)
> 
> I can see some ongoing [issue](https://github.com/Azure/sonic-mgmt/issues/1699) related to this matter? Any Idea @daall?

We can re-run the tests. It looks like everything else is working now so hopefully it passes on the next run. :)

**Comment by renukamanavalan** [comment]
From the failed build logs:
[2020-06-16T18:40:00.200Z] nss_tacplus.c: In function 'parse_config':
[2020-06-16T18:40:00.200Z] nss_tacplus.c:251:5: error: this 'if' clause does not guard... [-Werror=misleading-indentation]
[2020-06-16T18:40:00.200Z]      if(source_addr)
[2020-06-16T18:40:00.200Z]      ^~
[2020-06-16T18:40:00.200Z] nss_tacplus.c:253:9: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the 'if'
[2020-06-16T18:40:00.200Z]          source_addr = NULL;
[2020-06-16T18:40:00.200Z]          ^~~~~~~~~~~
